torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,143 @@
+                           )
+
+COMPLETE_DTYPES_DISPATCH = (
+    all_types,
+    all_types_and_complex,
+    all_types_and_half,
+    floating_types,
+    floating_and_complex_types,
+    floating_types_and_half,
+    integral_types,
+    complex_types,
+)
+
+EXTENSIBLE_DTYPE_DISPATCH = (
+    all_types_and_complex_and,
+    floating_types_and,
+    floating_and_complex_types_and,
+    integral_types_and,
+    all_types_and,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+class _dynamic_dispatch_dtypes(_dispatch_dtypes):
+    # Class to tag the dynamically generated types.
+    pass
+
+
+def get_supported_dtypes(op, sample_inputs_fn, device_type):
+    # Returns the supported dtypes for the given operator and device_type pair.
+    assert device_type in ['cpu', 'cuda']
+    if not TEST_CUDA and device_type == 'cuda':
+        warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA dtypes could be wrong")
+        return _dynamic_dispatch_dtypes(())
+
+    supported_dtypes = set()
+    for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
+        try:
+            samples = sample_inputs_fn(op, device_type, dtype, False)
+        except RuntimeError:
+            # If `sample_inputs_fn` doesn't support sampling for a given
+            # `dtype`, we assume that the `dtype` is not supported.
+            continue
+
+        # We assume the dtype is supported
+        # only if all samples pass for the given dtype.
+        supported = True
+        for sample in samples:
+            try:
+                op(sample.input, *sample.args, **sample.kwargs)
+            except RuntimeError as re:
+                if "not implemented for" not in str(re):
+                    raise re
+                else:
+                    # dtype is not supported
+                    supported = False
+                    break
+
+        if supported:
+            supported_dtypes.add(dtype)
+
+    return _dynamic_dispatch_dtypes(supported_dtypes)
+
+
+def dtypes_dispatch_hint(dtypes):
+    # Function returns the appropriate dispatch function (from COMPLETE_DTYPES_DISPATCH and EXTENSIBLE_DTYPE_DISPATCH)
+    # and its string representation for the passed `dtypes`.
+    return_type = collections.namedtuple('return_type', 'dispatch_fn dispatch_fn_str')
+
+    # CUDA is not available, dtypes will be empty.
+    if len(dtypes) == 0:
+        return return_type((), str(tuple()))
+
+    set_dtypes = set(dtypes)
+    for dispatch in COMPLETE_DTYPES_DISPATCH:
+        # Short circuit if we get an exact match.
+        if set(dispatch()) == set_dtypes:
+            return return_type(dispatch, dispatch.__name__ + "()")
+
+    # Metric for finding the closest dispatcher
+    # Intersection Over Union
+    def iou(dispatch_dtypes):
+        return len(set_dtypes.intersection(dispatch_dtypes)) * 1.0 / len(set_dtypes.union(dispatch_dtypes))
```

Comment:
because `dispatch_dtypes` must now be a subset of `set_dtypes` this equation reduces to:

```
len(dispatch_dtypes) / len(set_dtypes)
```

When comparing these values the denominator of each of them is fixed at `len(set_dtypes)`, so this is equivalent to just comparing `len(dispatch_dtypes)`, which would be more readonable.

Reply:
Aha! Thanks for the explanation :)!